src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs

```diff
@@ -243,67 +207,74 @@ internal int RemoveExpiredValues()
             return numItemsRemoved;
         }
 
-        /// <summary>
-        /// The timer callback that adds a request to remove expired items from the event queue.
-        /// </summary>
-        /// <param name="state">the timer state</param>
-        protected void RemoveExpiredValuesPeriodically(object state)
-        {
-            if (_map.Count > 0)
-                AddActionToEventQueue(() => RemoveExpiredValues());
-        }
-
         /// <summary>
         /// Remove items from the LinkedList by the desired compaction percentage.
         /// This should be a private method.
         /// </summary>
         private void RemoveLRUs()
         {
             // use the _capacity for the newCacheSize calculation in the case where the cache is experiencing overflow
-            int currentCount = _map.Count <= _capacity ? _capacity : _map.Count;
-            var newCacheSize = currentCount - (int)(currentCount * _compactionPercentage);
-            while (_map.Count > newCacheSize && _doubleLinkedList.Count > 0)
+            var newCacheSize = CalculateNewCacheSize();
+            while (base.Count > newCacheSize && _doubleLinkedList.Count > 0)
             {
                 var lru = _doubleLinkedList.Last;
-                if (_map.TryRemove(lru.Value.Key, out var cacheItem))
+                if (base.TryRemoveInternal(lru.Value.Key, out var cacheItem))
                     OnItemRemoved?.Invoke(cacheItem.Value);
 
                 _doubleLinkedList.Remove(lru);
             }
         }
 
         /// <summary>
-        /// This is the method that determines the end time for the event queue task.
-        /// The goal is to be able to track the incoming events and predict how long the task should run in order to
-        /// avoid a long running task and reduce the overhead costs of restarting tasks.
-        /// For example, maybe we can track the last three events' time and set the _eventQueueRunDurationInSeconds = 2 * average_time_between_events.
-        /// Note: tasks are based on thread pool so the overhead should not be huge but we should still try to minimize it.
+        /// This method is called after an item is added to the event queue, and the event queue task needs to be started if not running (_eventQueueTaskState != EventQueueTaskRunning).
+        /// Setting _eventQueueTaskState to EventQueueTaskRunning should prevent multiple task being started.
         /// </summary>
-        /// <returns>the time when the event queue task should end</returns>
-        private DateTime SetTaskEndTime()
+        private void StartEventQueueTaskIfNotRunning()
         {
-            return DateTime.UtcNow.AddSeconds(EventQueueTaskIdleTimeoutInSeconds);
+            // Setting _eventQueueTaskState to EventQueueTaskDoNotStop here will force the event queue task in EventQueueTaskAction to continue even it has past the _eventQueueTaskEndTime.
```

Comment:
_eventQueueTaskEndTime doesn't exist anymore, but it's referenced in few different comments.